### PR TITLE
fix: change HPA minReplicas to 2 to resolve PDB issues

### LINKED ISSUE #12602: unable to drain k8s node running Istio components with PDB
Deployed istio 1.1.0 RC with helm chart (not template)

I would like to drain out a specific node because it needs a reboot. 
The PODs
```
Annotations:        scheduler.alpha.kubernetes.io/critical-pod:
```
prevents that.

```
error when evicting pod "istio-policy-d48655744-s8ggt" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

(also note that annotation scheduler.alpha.kubernetes.io/critical-pod is deprecated and will be removed in k8s 1.14!)

### LINKED ISSUE #24000: Validate PDB and HPA settings are compatible
autoscaling enabled causes PDBs to be installed, but there's no enforcement of minReplicas to be compliant with PDB which means that if the value is too low pods can be completely prevented from restarting, blocking rescheduling, upgrade etc. 
This feature would at least validate that the settings are correct.

### PR DESCRIPTION
**Pull request by agilgur5**
> Please provide a description for what this PR is for.

## Description

- PDBs for Istio services have a [minAvailable of 1](https://github.com/istio/istio/blob/c2b87ce01626d1dac3b4c7ff33fa83bb7ffc945b/manifests/charts/gateways/istio-ingress/templates/poddisruptionbudget.yaml#L12), and HPAs had a
  minReplicas of 1 (as well as default replicaCount of 1), meaning that
  the defaults would cause failures for things like kops rolling-update

- bump minReplicas to 2 so that at least 1 pod can go down by default
  - and bump autoScaleMin to 2 as well, which [seems to correspond to
    HPA minReplicas](https://github.com/istio/istio/blob/a671e62bd1ca3e6fd09bd0fbe03545c136b0f3f5/manifests/charts/gateways/istio-ingress/templates/autoscale.yaml#L13) as well
  - and bump default replicaCount to 2 where it was previously 1
- ~also bump maxReplicas to 10 to account for the doubling of minReplicas~
  - ~and autoScaleMax respectively~ **EDIT**: removed according to review

## Tags

Fixes #24000 . _EDIT_: Also fixes #12602 which duplicates it. Or perhaps the former isn't fully resolved without adding a validation piece.

> And to help us figure out who should review this PR, please 
> put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

This changes the _defaults_, so maybe it affects "Installation" too? And it does change (only) Configuration, but not "Configuration Infrastructure". Let me know if I should check more, I was not really sure and these areas weren't described in the contributing guidelines.

## Review Notes

Hi there, first-time contributor here 👋  I was affected by #24000 when performing rolling-updates and saw it was marked as "good first issue", so thought I would try my hand at fixing it. Please let me know if any changes are necessary and offer any guidance 🙏 

Things I wasn't totally sure of:
1. Per the description, I added `autoScaleMin` as I saw it was used in several places and `minReplicas` was set to it, but let me know if that's wrong.

1. ~I doubled `maxReplicas` / `autoScaleMax` to correspond to the doubled minimum, similar to [some sample code](https://github.com/istio/istio/blob/d3f224cdfb4ca2c3f813b1cfa3d68dfe0a05aaa3/operator/samples/pilot-k8s.yaml#L12), but if this should be left unchanged, can undo it.~ **EDIT** removed according to review
1. I think `replicaCount` needs to be set to `2` in more places and perhaps in some places. I only changed the ones that were previously set for safety, but can add more. May need guidance on that.
1. Not really sure how to test this. Ideally would compare PDB `minAvailable` to default `replicaCount` and `minReplicas`, but those can be in lots of different places and I make no claims I understand how the multitudes of configs here work 😅 If someone could tell me where to look / guide me in the right direction, that would be helpful!

### CODE REVIEW

**Comment by agilgur5** [comment]
@googlebot I signed it

**Review by morvencao — COMMENTED** [review]
Need to add validation logic to resolve issue, not only change the default value for PDB and HPA.
End users may configure incompliant minReplica value for PDB with HPA which prevents pods from restarting, blocking rescheduling, upgrade...

**Comment by sdake** [comment]
Please read through https://github.com/istio/istio/issues/12602 - there is a pretty good analysis of the problem.

File: manifests/charts/gateways/istio-egress/values.yaml [hunk 1]
```diff
@@ -31,12 +31,12 @@ gateways:
...
     # Scalability tuning
-    # replicaCount: 1
+    # replicaCount: 2
     rollingMaxSurge: 100%
     rollingMaxUnavailable: 25%
     autoscaleEnabled: true
-    autoscaleMin: 1
-    autoscaleMax: 5
+    autoscaleMin: 2
+    autoscaleMax: 10
```
Comment (sdake) [inline manifests/charts/gateways/istio-egress/values.yaml:39]: changing the max may have unexpected consequences.

File: manifests/profiles/default.yaml [hunk 2]
```diff
@@ -68,8 +68,8 @@ spec:
               targetPort: 15443
               name: tls
         hpaSpec:
-          maxReplicas: 5
-          minReplicas: 1
+          maxReplicas: 10
```
Comment (sdake) [inline manifests/profiles/default.yaml:71]: I don't think there is a good reason to change maxReplicas, especially here.

File: manifests/profiles/default.yaml [hunk 3]
```diff
@@ -232,9 +232,9 @@ spec:
       validationURL: ""
     pilot:
       autoscaleEnabled: true
-      autoscaleMin: 1
-      autoscaleMax: 5
-      replicaCount: 1
+      autoscaleMin: 2
```
Comment (sdake) [inline manifests/profiles/default.yaml:235]: `autoscaleMin: 1` is not sufficient to drain a node? Does K8s generate a validation error?

**Review by sdake — COMMENTED** [review]
As an initial PR, I think it is fine not to validate the defaults. More important is to fix this problem and backport this for 1.7.

I don't understand why the hpa spec replicas are increased to 2. Seems like it would be sufficient to keep hpa spec at 1/5, and then increase the replicas to 2.

**Comment by sdake** [comment]
@agilgur5 nice first PR! Thank you for the contribution. We have struggled and struggled a project with what to do about HPAs. The default for Istio is to install HPAs, therefore, the defaults should account for HPAs being installed. After reading the latest K8s docs on ratios used to determine an HPA, I am not clear this PR is correct, but you are on the right track.

Cheers,
-steve

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't feel comfortable making the min replicas==1 for the default profile, as this will force all users, no matter how small of a cluster, to have double the resources used which is often a waste. Its also harder to debug, etc 2 replicas. I would rather have:

* no PDB and have a "HA" profile with PDB+min replicas=2+
* Kubernetes change to allow scaling up to meet the constraints of a PDB

**Comment by sdake** [comment]
> I don't feel comfortable making the min replicas==1 for the default profile, as this will force all users, no matter how small of a cluster, to have double the resources used which is often a waste. Its also harder to debug, etc 2 replicas. I would rather have:
> 
> * no PDB and have a "HA" profile with PDB+min replicas=2+
> * Kubernetes change to allow scaling up to meet the constraints of a PDB

Agree. It is a shame min replicas can not be zero unless alpha gates are turned on in K8s 1.18. That would be an easier answer than adding a profile. We do need to take care not to introduce profile proliferation. 👍 

But turning the suggestion around, wouldn't it be better for the default profile to include HA, and a different profile to disable PDBs? We could add a notice that the default profile includes Xyz features, including HA capability to the profile - to be displayed by istioctl.

@esnible thoughts?

Cheers,
-steve

**Comment by agilgur5** [comment]
> We have struggled and struggled a project with what to do about HPAs. The default for Istio is to install HPAs, therefore, the defaults should account for HPAs being installed.

> But turning the suggestion around, wouldn't it be better for the default profile to include HA, and a different profile to disable PDBs? We could add a notice that the default profile includes Xyz features, including HA capability

Throwing in my two cents, my response to both of these would be that I generally think of security, availability matters as things you have turned on by default -- best practices for production use turned on by default. One could also describe that as "making the hard things easy", there's then little effort in achieving the best practice. Then if there is a need, have a very **explicit** _opt-out_ capability for PoCs or smaller clusters etc. Making security or HA _opt-in_ allows "misconfigurations" and shortcuts to proliferate in my experience. It requires _effort_ (and knowledge of their existence too) to use best practices and therefore it can often be forgotten or ignored.
Perhaps worthwhile to note, the decision also impacts the way one thinks of things, either as maintainer or as user. In opt-out, HA is the default and "forefront" with non-HA the "exception", whereas opt-in, HA is the "exception" and non-HA is forefront.

> We do need to take care not to introduce profile proliferation. 👍

I think this is relevant too, as the current defaults already include a PDB together with a misconfigured default HPA (that this PR attempts to fix). Also, there is the existing `minimal` profile that I would think could be used for a non-HA set-up as well.

<hr>

Y'all obviously understand Istio significantly better than me, but thought I'd throw in my two cents too as a k8s admin and user (as well as a library author and contributor myself). I occasionally find myself searching for HA and secure configuration in various charts / components and finding them missing because the defaults lack them -- adding them is considerable work and even if they exist they may not be as well-tested as the defaults. But best practices shouldn't be after thoughts (not that the Istio team treats them as such), especially for components that have wide impact on the cluster.

File: manifests/profiles/default.yaml (hunk 2 above)
Comment (agilgur5) [reply manifests/profiles/default.yaml:71]: Ok, I can remove the changes to `maxReplicas` everywhere then. Per my review notes, I wasn't sure if that was necessarily warranted.

**Comment by agilgur5** [comment]
> Need to add validation logic to resolve issue, not only change the default value for PDB and HPA.

@morvencao would be happy to try adding that (perhaps in a separate PR so as to not block this) if you could point me in the right direction of where that might go. Per my review notes, I don't really know where to look for something like that or tests.

> I don't understand why the hpa spec replicas are increased to 2. Seems like it would be sufficient to keep hpa spec at 1/5, and then increase the replicas to 2.

The issue(s) pointed to changing this to 2, but I do think it is necessary myself as well. If only 1 replica is available (whether that's set by HPA or the default `replicaCount` isn't relevant as I understand), the PDB will result in an error response to an eviction request. The HPA only scales up with response to utilization, not a pending eviction. But correct me if I'm wrong, I'm certainly not an expert in this.

**Comment by howardjohn** [comment]
I opened https://github.com/kubernetes/kubernetes/issues/93476 about this
on k8s side. I don't think we should just wait on k8s, but I am pretty
convinced this is poor (but expected) behavior on k8s side.

**Comment by agilgur5** [comment]
On the k8s UX side, I actually thought a rolling update strategy might apply to the PDB case too (that similarly means a user is ok with some level of scale up), but that's only for the resource's rolling update itself, not if the underlying nodes go through a rolling update. All of these relate to each other, but are independent without knowledge of each other on the k8s side I guess.

**Comment by rcernich** [comment]
I wonder if a better solution is to configure `maxSurge` on the deployments.

**Comment by sdake** [comment]
> I wonder if a better solution is to configure `maxSurge` on the deployments.

@rcernich Reading https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#max-surge - I don't immediately understand your proposal. The default (according to that page) is 25%. This means the max surge during a rolling update with our defaults of 5 max pods = 6 pods. This should enable 1 extra pod for a rolling update. Not sure why we don't see this behavior automatically?

I still believe we have two choices:
* Add HA to the default profile
- or -
* Add an HA profile that is essentially the default profile (perhaps default-ha)

I think default-ha seems like a reasonable approach, although as a project we do not want a HA profile version for every profile in the system.

An entirely different option is to have a conditional set of HA flags such as

`istioctl --set high_availability=true` or the like.

@ostromart @costinm @esnible @howardjohn  can you comment on the above. I think we have punted on this problem for about three years, and its time to take a hard look at solving it.

Cheers,
-steve

**Comment by howardjohn** [comment]
> I wonder if a better solution is to configure maxSurge on the deployments.

I 100% agree but I don't think k8s does this: https://github.com/kubernetes/kubernetes/issues/93476. If the issue is invalid and we can do it, please let us know, that would be perfect.


I don't see why we need a profile to configure a single flag. You can just set `values.pilot.replicas=5` or whatever the option is?

File: manifests/profiles/default.yaml (hunk 2 above)
Comment (agilgur5) [reply manifests/profiles/default.yaml]: Done

File: manifests/profiles/default.yaml (hunk 3 above)
Comment (agilgur5) [reply manifests/profiles/default.yaml:235]: Well if it scaled back down to 1, you'd have only 1 replica and therein again run into the PDB issue

**Comment by agilgur5** [comment]
> > I wonder if a better solution is to configure `maxSurge` on the deployments.
> 
> I 100% agree but I don't think k8s does this: [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476). If the issue is invalid and we can do it, please let us know, that would be perfect.

Yes, per my previous comment (https://github.com/istio/istio/pull/25873#issuecomment-664533507) it doesn't currently do this, `maxSurge` is strictly a deploy/`rollingUpdate` configuration, it does not affect the rolling of actual k8s nodes. That is perhaps out-of-scope with what `maxSurge` is meant to handle.

There does not seem to be much bite on the k8s issue filed https://github.com/kubernetes/kubernetes/issues/93476. Not that I'm a maintainer, but that too seems to perhaps be out-of-scope; HPA is then scaling not for load as it is purposed, but to meet PDB requirements, which also couples the two resources together.

> I don't see why we need a profile to configure a single flag. You can just set `values.pilot.replicas=5` or whatever the option is?

It's for several services -- I experienced the PDB error with both `istiod` and `istiocoredns`. And one needs to configure _both_ the `replicaCount` and the HPA `minReplicas`.

> I still believe we have two choices:
> 
> * Add HA to the default profile
> 
> * or -
> 
> * Add an HA profile that is essentially the default profile (perhaps default-ha)

Per my previous comment (https://github.com/istio/istio/pull/25873#issuecomment-664501966), if the goal is to cater to people "trying Istio out", the existing `minimal` or `demo` profiles I believe _already suffice_ and were built for this. I'd also argue that's not the primary use-case by definition and that someone trying it out may want to see HA out-of-the-box, or at least not experience bugs such as this.
For "default production usage", I think HA makes sense to have out-of-the-box, as it's production. If one makes the argument that HA requires too much resources, one could take a slippery slope and say the same for security -- but Istio does have a variety of security features out-of-the-box. A lack of either can be considered a misconfiguration and can be exploited.

In any case, the defaults are still currently incorrect and a confusing issue that lots of users run into (including those trying it out).

I've fixed the one review comment about `maxReplicas` which I had asked for guidance on, but have otherwise not received any guidance on this in months. As such this has been deadlocked while the bug continues to exist.

**Comment by agilgur5** [comment]
hit this bug again a year later with Istio 1.10 with a different cluster at a different job, so this _is_ still a bug... 😕

**Comment by marshallford** [comment]
I still have `pilot.autoscaleMin` set to `2` in my istiod helm chart values for Istio 1.17. I feel your pain @agilgur5.

**Pull request closed without merging**